File: Model/NetElems/Activation.fs
module Synapses.Model.NetElems.Activation

open System
open System.Text.Json.Serialization

type Function = { name: string
                  f: float -> float
                  deriv: float -> float
                  inverse: float -> float }

[<JsonFSharpConverter>]
type SerializableFunction = string

exception UnknownActivationF of string

let safeSigmoid (x: float): float =
    -x
    |> min 500.0
    |> Math.Exp
    |> (+) 1.0
    |> (/) 1.0

let sigmoid = { name = "sigmoid"
                f = safeSigmoid
                deriv = fun x ->
                    safeSigmoid x * (1.0 - safeSigmoid x)
                inverse = fun x ->
                    Math.Log(x / (1.0 - x)) }

let identity = { name = "identity"
                 f = fun x -> x
                 deriv = fun _ -> 1.0
                 inverse = fun x -> x }

let tanh = { name = "tanh"
             f = Math.Tanh
             deriv = fun x ->
                 1.0 - (Math.Tanh x) * (Math.Tanh x)
             inverse = fun x ->
                 0.5 * Math.Log((1.0 + x) / (1.0 - x)) }

let leakyReLU = { name = "leakyReLU"
                  f = fun x ->
                      match x with
                      | _ when x < 0.0 -> 0.01 * x
                      | _ -> x
                  deriv = fun x ->
                      match x with
                      | _ when x < 0.0 -> 0.01
                      | _ -> 1.0
                  inverse = fun x ->
                      match x with
                      | _ when x < 0.0 -> x / 0.01
                      | _ -> x }

let serialized
        (activF: Function):
        SerializableFunction =
    activF.name

let deserialized
    (s: SerializableFunction)
    : Function =
    match s with
    | "sigmoid" -> sigmoid
    | "identity" -> identity
    | "tanh" -> tanh
    | "leakyReLU" -> leakyReLU
    | _ -> raise (UnknownActivationF(s))

File: Scala/src/main/scala/synapses/model/netElems/Activation.scala
package synapses.model.netElems

case class Activation(name: String,
                      f: Double => Double,
                      deriv: Double => Double,
                      inverse: Double => Double)

object Activation {

  private def safeSigmoid(x: Double): Double = {
    val boundedPow = Math.min(-x, 500.0)
    1.0 / (1.0 + Math.exp(boundedPow))
  }

  val sigmoid: Activation =
    Activation(
      "sigmoid",
      safeSigmoid,
      x => safeSigmoid(x) * (1.0 - safeSigmoid(x)),
      x => Math.log(x / (1.0 - x))
    )

  val identity: Activation =
    Activation(
      "identity",
      x => x,
      _ => 1.0,
      x => x
    )

  val tanh: Activation =
    Activation(
      "tanh",
      Math.tanh,
      x => 1.0 - Math.tanh(x) * Math.tanh(x),
      x => 0.5 * Math.log((1.0 + x) / (1.0 - x))
    )

  val leakyReLU: Activation =
    Activation(
      "leakyReLU",
      x => if (x < 0.0) {
        0.01 * x
      } else {
        x
      },
      x => if (x < 0.0) {
        0.01
      } else {
        1.0
      },
      x => if (x < 0.0) {
        x / 0.01
      } else {
        x
      }
    )

  type ActivationSerialized = String

  def serialized(activation: Activation)
  : ActivationSerialized = activation.name

  def deserialized(activationSerialized: ActivationSerialized)
  : Activation =
    activationSerialized match {
      case "sigmoid" => Activation.sigmoid
      case "identity" => Activation.identity
      case "tanh" => Activation.tanh
      case "leakyReLU" => Activation.leakyReLU
    }

}
